src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,13 +620,144 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    // If there's any replications pending, wait for them to complete.
+    // Note that we throw away the result and use Raft's committed offset
+    // below. This deals with both success and failure paths.
+
+    if (_last_replicate) {
+        auto fut = std::exchange(_last_replicate, std::nullopt).value();
+        if (!fut.available()) {
+            vlog(
+              _logger.debug, "Waiting for ongoing replication before syncing");
+
+            try {
+                const auto before = model::timeout_clock::now();
+
+                std::ignore = co_await ss::with_timeout(
+                  before + timeout, std::move(fut));
+
+                const auto after = model::timeout_clock::now();
+
+                // Update the timeout whille accounting for under/overflow.
+                const auto duration = after > before
+                                        ? after - before
+                                        : ss::lowres_clock::duration{0};
+
+                if (duration >= timeout) {
+                    vlog(
+                      _logger.error,
+                      "Replication wait for archival STM timed out");
+                    co_return false;
+                }
+
+                timeout -= duration;
+            } catch (const ss::timed_out_error&) {
+                vlog(
+                  _logger.error, "Replication wait for archival STM timed out");
+                co_return false;
+            } catch (...) {
+                vlog(
+                  _logger.debug,
+                  "Replication failed for archival STM command. Continuing "
+                  "sync.");
```

Comment:
Why is it OK to continue? If we received some error from raft does this mean that we lost leadership and need to just exit?

Reply:
I don't think it really matters what we do on the replication failed case. If `replicate` returns an exceptional future, that should propagate and trigger the upload loop to check terms again. If `replicate` fails with an error code, we step and the rest of this function is a no-op.

Still, maybe it's easier to reason about if we return a sync failure in these cases. I'll do that.

Reply:
Changed it to exit early and added another test which simulates a replication failure.